File: rnd_warehouse_management/warehouse_management/page/approval_dashboard/approval_dashboard.js
frappe.pages['approval-dashboard'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Approval Dashboard',
		single_column: true
	});
	
	wrapper.approval_dashboard = new ApprovalDashboard(wrapper, page);
}

class ApprovalDashboard {
	constructor(wrapper, page) {
		this.wrapper = wrapper;
		this.page = page;
		this.filters = {};
		this.selected_entries = [];
		
		this.setup_page();
		this.load_data();
	}
	
	setup_page() {
		// Add refresh button
		this.page.set_primary_action(__('Refresh'), () => {
			this.load_data();
		}, 'fa fa-refresh');
		
		// Add bulk approve button
		this.page.add_action_icon('fa fa-check-circle', () => {
			this.bulk_approve_action();
		}, __('Bulk Approve'));
		
		// Setup filters
		this.setup_filters();
		
		// Create tabs
		this.create_tabs();
	}
	
	setup_filters() {
		// Movement Type filter
		this.page.add_field({
			fieldname: 'movement_type',
			label: __('Movement Type'),
			fieldtype: 'Link',
			options: 'Movement Type Master',
			change: () => {
				this.filters.movement_type = this.page.fields_dict.movement_type.get_value();
				this.load_data();
			}
		});
		
		// Date range filter
		this.page.add_field({
			fieldname: 'from_date',
			label: __('From Date'),
			fieldtype: 'Date',
			default: frappe.datetime.add_days(frappe.datetime.get_today(), -30),
			change: () => {
				this.filters.from_date = this.page.fields_dict.from_date.get_value();
				this.load_data();
			}
		});
		
		this.page.add_field({
			fieldname: 'to_date',
			label: __('To Date'),
			fieldtype: 'Date',
			default: frappe.datetime.get_today(),
			change: () => {
				this.filters.to_date = this.page.fields_dict.to_date.get_value();
				this.load_data();
			}
		});
	}
	
	create_tabs() {
		const $page_content = $(this.wrapper).find('.page-content');
		$page_content.html(`
			<div class="dashboard-container">
				<!-- Statistics Cards -->
				<div class="stats-section" id="stats-section"></div>
				
				<!-- Tabs -->
				<ul class="nav nav-tabs" role="tablist">
					<li class="nav-item">
						<a class="nav-link active" data-toggle="tab" href="#pending-tab" role="tab">
							<i class="fa fa-clock-o"></i> ${__('Pending Approvals')}
						</a>
					</li>
					<li class="nav-item">
						<a class="nav-link" data-toggle="tab" href="#history-tab" role="tab">
							<i class="fa fa-history"></i> ${__('My History')}
						</a>
					</li>
				</ul>
				
				<!-- Tab Content -->
				<div class="tab-content">
					<div class="tab-pane active" id="pending-tab" role="tabpanel">
						<div id="pending-approvals-list"></div>
					</div>
					<div class="tab-pane" id="history-tab" role="tabpanel">
						<div id="approval-history-list"></div>
					</div>
				</div>
			</div>
		`);
	}
	
	load_data() {
		frappe.call({
			method: 'rnd_warehouse_management.warehouse_management.page.approval_dashboard.approval_dashboard.get_dashboard_data',
			args: { filters: this.filters },
			callback: (r) => {
				if (r.message) {
					this.render_statistics(r.message.statistics);
					this.render_pending_approvals(r.message.pending_approvals);
					this.render_approval_history(r.message.approval_history);
				}
			}
		});
	}
	
	render_statistics(stats) {
		const $stats = $(this.wrapper).find('#stats-section');
		$stats.html(`
			<div class="row" style="margin-bottom: 20px;">
				<div class="col-sm-3">
					<div class="stat-card stat-pending">
						<div class="stat-icon"><i class="fa fa-clock-o"></i></div>
						<div class="stat-content">
							<h3>${stats.pending_count || 0}</h3>
							<p>${__('Pending Approvals')}</p>
						</div>
					</div>
				</div>
				<div class="col-sm-3">
					<div class="stat-card stat-approved">
						<div class="stat-icon"><i class="fa fa-check-circle"></i></div>
						<div class="stat-content">
							<h3>${stats.approved_this_month || 0}</h3>
							<p>${__('Approved This Month')}</p>
						</div>
					</div>
				</div>
				<div class="col-sm-3">
					<div class="stat-card stat-rejected">
						<div class="stat-icon"><i class="fa fa-times-circle"></i></div>
						<div class="stat-content">
							<h3>${stats.rejected_this_month || 0}</h3>
							<p>${__('Rejected This Month')}</p>
						</div>
					</div>
				</div>
				<div class="col-sm-3">
					<div class="stat-card stat-overdue">
						<div class="stat-icon"><i class="fa fa-exclamation-triangle"></i></div>
						<div class="stat-content">
							<h3>${stats.overdue_count || 0}</h3>
							<p>${__('Overdue (>3 days)')}</p>
						</div>
					</div>
				</div>
			</div>
		`);
	}
	
	render_pending_approvals(approvals) {
		const $list = $(this.wrapper).find('#pending-approvals-list');
		
		if (!approvals || approvals.length === 0) {
			$list.html(`
				<div class="empty-state">
					<i class="fa fa-check-circle" style="font-size: 48px; color: #28a745;"></i>
					<p>${__('No pending approvals')}</p>
				</div>
			`);
			return;
		}
		
		let html = `
			<div class="approval-actions" style="margin: 15px 0;">
				<button class="btn btn-xs btn-primary" id="select-all-btn">${__('Select All')}</button>
				<button class="btn btn-xs btn-default" id="deselect-all-btn">${__('Deselect All')}</button>
			</div>
			<div class="approval-list">
		`;
		
		approvals.forEach((entry) => {
			const urgency_class = entry.urgency === 'High' ? 'urgency-high' : 
								  entry.urgency === 'Medium' ? 'urgency-medium' : 'urgency-low';
			
			html += `
				<div class="approval-card ${urgency_class}" data-name="${entry.name}">
					<div class="approval-checkbox">
						<input type="checkbox" class="entry-checkbox" data-name="${entry.name}">
					</div>
					<div class="approval-content">
						<div class="approval-header">
							<h4><a href="/app/stock-entry/${entry.name}" target="_blank">${entry.name}</a></h4>
							<span class="badge badge-warning">${entry.custom_approval_status}</span>
						</div>
						<div class="approval-details">
							<div><strong>${__('Movement Type')}:</strong> ${entry.custom_sap_movement_type} - ${entry.custom_movement_type_description || ''}</div>
							<div><strong>${__('Posting Date')}:</strong> ${entry.posting_date}</div>
							<div><strong>${__('Days Pending')}:</strong> ${entry.days_pending} days 
								<span class="urgency-badge ${urgency_class}">${entry.urgency}</span>
							</div>
							<div><strong>${__('Created By')}:</strong> ${entry.owner}</div>
						</div>
						<div class="approval-actions">
							<button class="btn btn-sm btn-success approve-btn" data-name="${entry.name}">
								<i class="fa fa-check"></i> ${__('Approve')}
							</button>
							<button class="btn btn-sm btn-danger reject-btn" data-name="${entry.name}">
								<i class="fa fa-times"></i> ${__('Reject')}
							</button>
							<button class="btn btn-sm btn-default view-btn" data-name="${entry.name}">
								<i class="fa fa-eye"></i> ${__('View')}
							</button>
						</div>
					</div>
				</div>
			`;
		});
		
		html += '</div>';
		$list.html(html);
		
		// Attach event handlers
		this.attach_approval_handlers();
	}
	
	render_approval_history(history) {
		const $list = $(this.wrapper).find('#approval-history-list');
		
		if (!history || history.length === 0) {
			$list.html(`
				<div class="empty-state">
					<i class="fa fa-history" style="font-size: 48px; color: #6c757d;"></i>
					<p>${__('No approval history')}</p>
				</div>
			`);
			return;
		}
		
		let html = `
			<table class="table table-bordered table-hover">
				<thead>
					<tr>
						<th>${__('Date/Time')}</th>
						<th>${__('Stock Entry')}</th>
						<th>${__('Movement Type')}</th>
						<th>${__('Action')}</th>
						<th>${__('Level')}</th>
						<th>${__('Status')}</th>
						<th>${__('Comments')}</th>
					</tr>
				</thead>
				<tbody>
		`;
		
		history.forEach((log) => {
			const status_class = log.approval_status === 'Approved' ? 'text-success' : 
								 log.approval_status === 'Rejected' ? 'text-danger' : 'text-warning';
			
			html += `
				<tr>
					<td>${log.event_timestamp || ''}</td>
					<td><a href="/app/stock-entry/${log.stock_entry}" target="_blank">${log.stock_entry}</a></td>
					<td>${log.movement_type || ''} - ${log.movement_description || ''}</td>
					<td>${log.event_type || ''}</td>
					<td>${log.approval_level || 'N/A'}</td>
					<td class="${status_class}"><strong>${log.approval_status || ''}</strong></td>
					<td>${log.comments || ''}</td>
				</tr>
			`;
		});
		
		html += `
				</tbody>
			</table>
		`;
		$list.html(html);
	}
	
	attach_approval_handlers() {
		const me = this;
		
		// Select/Deselect all
		$(me.wrapper).find('#select-all-btn').on('click', function() {
			$(me.wrapper).find('.entry-checkbox').prop('checked', true);
		});
		
		$(me.wrapper).find('#deselect-all-btn').on('click', function() {
			$(me.wrapper).find('.entry-checkbox').prop('checked', false);
		});
		
		// Approve button
		$(me.wrapper).find('.approve-btn').on('click', function() {
			const entry_name = $(this).data('name');
			me.approve_entry(entry_name);
		});
		
		// Reject button
		$(me.wrapper).find('.reject-btn').on('click', function() {
			const entry_name = $(this).data('name');
			me.reject_entry(entry_name);
		});
		
		// View button
		$(me.wrapper).find('.view-btn').on('click', function() {
			const entry_name = $(this).data('name');
			frappe.set_route('Form', 'Stock Entry', entry_name);
		});
	}
	
	approve_entry(entry_name) {
		const me = this;
		
		const dialog = new frappe.ui.Dialog({
			title: __('Approve Stock Entry'),
			fields: [
				{
					fieldname: 'comments',
					fieldtype: 'Small Text',
					label: __('Approval Comments')
				}
			],
			primary_action_label: __('Approve'),
			primary_action: function(values) {
				frappe.call({
					method: 'rnd_warehouse_management.warehouse_management.stock_entry.approve_stock_entry',
					args: {
						stock_entry_name: entry_name,
						comments: values.comments
					},
					callback: function(r) {
						if (r.message && r.message.success) {
							frappe.show_alert({ message: r.message.message, indicator: 'green' }, 5);
							dialog.hide();
							me.load_data();
						}
					}
				});
			}
		});
		
		dialog.show();
	}
	
	reject_entry(entry_name) {
		const me = this;
		
		const dialog = new frappe.ui.Dialog({
			title: __('Reject Stock Entry'),
			fields: [
				{
					fieldname: 'rejection_reason',
					fieldtype: 'Small Text',
					label: __('Rejection Reason'),
					reqd: 1
				}
			],
			primary_action_label: __('Reject'),
			primary_action: function(values) {
				frappe.call({
					method: 'rnd_warehouse_management.warehouse_management.stock_entry.reject_stock_entry',
					args: {
						stock_entry_name: entry_name,
						rejection_reason: values.rejection_reason
					},
					callback: function(r) {
						if (r.message && r.message.success) {
							frappe.show_alert({ message: r.message.message, indicator: 'orange' }, 5);
							dialog.hide();
							me.load_data();
						}
					}
				});
			}
		});
		
		dialog.show();
	}
	
	bulk_approve_action() {
		const me = this;
		const selected = [];
		
		$(me.wrapper).find('.entry-checkbox:checked').each(function() {
			selected.push($(this).data('name'));
		});
		
		if (selected.length === 0) {
			frappe.msgprint(__('Please select at least one entry'));
			return;
		}
		
		frappe.confirm(
			__('Are you sure you want to approve {0} entries?', [selected.length]),
			function() {
				frappe.call({
					method: 'rnd_warehouse_management.warehouse_management.page.approval_dashboard.approval_dashboard.bulk_approve',
					args: {
						stock_entry_names: selected,
						comments: 'Bulk approved from dashboard'
					},
					callback: function(r) {
						if (r.message) {
							const results = r.message;
							frappe.msgprint({
								title: __('Bulk Approve Results'),
								indicator: 'green',
								message: __('Approved: {0}, Failed: {1}', [results.success.length, results.failed.length])
							});
							me.load_data();
						}
					}
				});
			}
		);
	}
}

File: rnd_warehouse_management/rnd_warehouse_management/page/approval_dashboard/approval_dashboard.py
import frappe
from frappe import _
from frappe.utils import nowdate, add_days, get_datetime

@frappe.whitelist()
def get_dashboard_data(filters=None):
	"""Get approval dashboard data"""
	if not filters:
		filters = {}
	
	user = frappe.session.user
	user_roles = frappe.get_roles(user)
	
	# Get pending approvals
	pending_approvals = get_pending_approvals_for_dashboard(user, filters)
	
	# Get approval history
	approval_history = get_my_approval_history(user, filters)
	
	# Get statistics
	stats = get_approval_statistics(user)
	
	return {
		"pending_approvals": pending_approvals,
		"approval_history": approval_history,
		"statistics": stats
	}


def get_pending_approvals_for_dashboard(user, filters):
	"""Get all pending approvals for the user"""
	from rnd_warehouse_management.warehouse_management.stock_entry import can_approve
	
	# Build filter conditions
	filter_conditions = {
		"custom_approval_status": ["like", "Pending%"],
		"docstatus": 0
	}
	
	# Add movement type filter
	if filters.get("movement_type"):
		filter_conditions["custom_sap_movement_type"] = filters["movement_type"]
	
	# Add date filter
	if filters.get("from_date"):
		filter_conditions["posting_date"] = [">=", filters["from_date"]]
	
	if filters.get("to_date"):
		filter_conditions["posting_date"] = ["<=", filters["to_date"]]
	
	# Get all pending stock entries
	pending_entries = frappe.get_all(
		"Stock Entry",
		filters=filter_conditions,
		fields=[
			"name", "posting_date", "custom_sap_movement_type", 
			"custom_movement_type_description", "custom_approval_status", 
			"custom_current_approval_level", "custom_pending_approver", 
			"custom_approval_requested_on", "owner", "creation"
		],
		order_by="creation desc"
	)
	
	# Filter by approval permission
	filtered_entries = []
	for entry in pending_entries:
		stock_entry = frappe.get_doc("Stock Entry", entry.name)
		if can_approve(stock_entry, user):
			# Calculate days pending
			if entry.get("custom_approval_requested_on"):
				days_pending = frappe.utils.date_diff(
					nowdate(),
					frappe.utils.getdate(entry.custom_approval_requested_on)
				)
				entry["days_pending"] = days_pending
				
				# Determine urgency
				if days_pending > 3:
					entry["urgency"] = "High"
				elif days_pending > 1:
					entry["urgency"] = "Medium"
				else:
					entry["urgency"] = "Low"
			else:
				entry["days_pending"] = 0
				entry["urgency"] = "Low"
			
			filtered_entries.append(entry)
	
	return filtered_entries


def get_my_approval_history(user, filters):
	"""Get approval history for the user"""
	# Build filter conditions
	filter_conditions = {
		"event_type": ["in", ["Approval Granted", "Approval Rejected"]],
		"user": user
	}
	
	# Add date filter
	if filters.get("from_date"):
		filter_conditions["event_timestamp"] = [">=", filters["from_date"] + " 00:00:00"]
	
	if filters.get("to_date"):
		filter_conditions["event_timestamp"] = ["<=", filters["to_date"] + " 23:59:59"]
	
	# Get audit logs
	audit_logs = frappe.get_all(
		"Stock Entry Audit Log",
		filters=filter_conditions,
		fields=[
			"stock_entry", "event_type", "event_timestamp", 
			"approval_level", "approval_status", "comments"
		],
		order_by="event_timestamp desc",
		limit=50
	)
	
	# Enrich with stock entry details
	for log in audit_logs:
		try:
			entry_data = frappe.db.get_value(
				"Stock Entry",
				log.stock_entry,
				["custom_sap_movement_type", "custom_movement_type_description"],
				as_dict=True
			)
			if entry_data:
				log["movement_type"] = entry_data.get("custom_sap_movement_type")
				log["movement_description"] = entry_data.get("custom_movement_type_description")
		except Exception:
			pass
	
	return audit_logs


def get_approval_statistics(user):
	"""Get approval statistics for the user"""
	user_roles = frappe.get_roles(user)
	
	# Count pending approvals
	pending_count = frappe.db.count(
		"Stock Entry",
		filters={
			"custom_approval_status": ["like", "Pending%"],
			"docstatus": 0
		}
	)
	
	# Count approvals granted this month
	first_day_of_month = frappe.utils.get_first_day(nowdate())
	approved_this_month = frappe.db.count(
		"Stock Entry Audit Log",
		filters={
			"user": user,
			"event_type": "Approval Granted",
			"event_timestamp": [">=", first_day_of_month + " 00:00:00"]
		}
	)
	
	# Count rejections this month
	rejected_this_month = frappe.db.count(
		"Stock Entry Audit Log",
		filters={
			"user": user,
			"event_type": "Approval Rejected",
			"event_timestamp": [">=", first_day_of_month + " 00:00:00"]
		}
	)
	
	# Count overdue approvals (>3 days pending)
	three_days_ago = add_days(nowdate(), -3)
	overdue_count = frappe.db.count(
		"Stock Entry",
		filters={
			"custom_approval_status": ["like", "Pending%"],
			"docstatus": 0,
			"custom_approval_requested_on": ["<=", three_days_ago + " 23:59:59"]
		}
	)
	
	return {
		"pending_count": pending_count,
		"approved_this_month": approved_this_month,
		"rejected_this_month": rejected_this_month,
		"overdue_count": overdue_count
	}


@frappe.whitelist()
def bulk_approve(stock_entry_names, comments=None):
	"""Bulk approve multiple stock entries"""
	if isinstance(stock_entry_names, str):
		import json
		stock_entry_names = json.loads(stock_entry_names)
	
	results = {
		"success": [],
		"failed": []
	}
	
	for entry_name in stock_entry_names:
		try:
			frappe.call(
				"rnd_warehouse_management.warehouse_management.stock_entry.approve_stock_entry",
				stock_entry_name=entry_name,
				comments=comments
			)
			results["success"].append(entry_name)
		except Exception as e:
			results["failed"].append({"name": entry_name, "error": str(e)})
	
	frappe.db.commit()
	
	return results


@frappe.whitelist()
def get_movement_types_for_filter():
	"""Get available movement types for filtering"""
	movement_types = frappe.get_all(
		"Movement Type Master",
		filters={"is_active": 1, "requires_approval": 1},
		fields=["movement_code", "description"],
		order_by="movement_code"
	)
	
	return movement_types
